utils/writeable/src/fallible.rs

```diff
@@ -0,0 +1,660 @@
+    }
+}
+
+/// Lower-level trait for writeables that can fail while writing.
+///
+/// The methods on this trait should be implemented on custom writeables, but they generally
+/// should not called directly. For methods more useful in client code, see [`TryWriteable`].
+///
+/// # Implementor Notes
+///
+/// All functions take a `handler` argument. This argument is used by [`TryWriteable`] to
+/// customize the error handling behavior.
+///
+/// When your writeable implementation encounters an error, call the handler with an instance
+/// of your error type. The handler will either return `Ok` with a [`Writeable`] that should be
+/// written to the sink, or `Err` with an error that should be immediately returned up the stack.
+///
+/// When in lossy mode, [`FallibleWriteable::Error`] is itself used as the writeable replacement.
+///
+/// # Examples
+///
+/// Implementing [`FallibleWriteable`] on a custom type:
+///
+/// ```
+/// use core::fmt;
+/// use writeable::assert_writeable_eq;
+/// use writeable::FallibleWriteable;
+/// use writeable::TryWriteable;
+/// use writeable::Writeable;
+///
+/// struct HelloWriteable<T> {
+///     name: Option<T>,
+/// }
+///
+/// impl<T: Writeable> FallibleWriteable for HelloWriteable<T> {
+///     type Error = &'static str;
+///
+///     fn fallible_write_to<
+///         W: fmt::Write + ?Sized,
+///         L: Writeable,
+///         E,
+///         F: FnMut(Self::Error) -> Result<L, E>,
+///     >(
+///         &self,
+///         sink: &mut W,
+///         mut handler: F,
+///     ) -> Result<Result<(), E>, fmt::Error> {
+///         sink.write_str("Hello, ")?;
+///         match self.name.as_ref() {
+///             Some(name) => name.write_to(sink)?,
+///             None => match handler("___") {
+///                 Ok(replacement) => replacement.write_to(sink)?,
+///                 Err(e) => return Ok(Err(e)),
+///             },
+///         };
+///         sink.write_char('!')?;
+///         Ok(Ok(()))
+///     }
+///
+///     // TODO: Implement the other methods in a similar way.
+/// }
+///
+/// // See it in action:
+/// assert_writeable_eq!(
+///     HelloWriteable {
+///         name: Some("Alice")
+///     }
+///     .lossy(),
+///     "Hello, Alice!"
+/// );
+/// assert_writeable_eq!(
+///     HelloWriteable { name: None::<&str> }.lossy(),
+///     "Hello, ___!"
+/// );
+/// ```
+pub trait FallibleWriteable {
+    /// The error type that the writeable may return.
+    ///
+    /// This type should implement [`Writeable`] so it can be used as a replacement in lossy mode.
+    type Error;
+
+    /// Writes a string to the given sink.
+    ///
+    /// This is a low-level function for implementors. For more details, see [`FallibleWriteable`].
+    fn fallible_write_to<
+        W: fmt::Write + ?Sized,
+        L: Writeable,
+        E,
+        F: FnMut(Self::Error) -> Result<L, E>,
+    >(
+        &self,
+        sink: &mut W,
+        handler: F,
+    ) -> Result<Result<(), E>, fmt::Error>;
```

Comment:
Discussion: This signature is nice in the sense that it enforces consistency while handling errors via the type system, but it also may cause more monomorphization. We'd probably need to add it to `write_pattern`. However, the creep should be limited to which version of TryWriteable the user calls.

```rust
pub(crate) fn write_pattern<'data, T, W, DS, TS>(
    pattern_items: impl Iterator<Item = PatternItem>,
    pattern_metadata: PatternMetadata,
    date_symbols: Option<&DS>,
    time_symbols: Option<&TS>,
    loc_datetime: &impl LocalizedDateTimeInput<T>,
    fixed_decimal_format: &FixedDecimalFormatter,
    w: &mut W,
) -> Result<(), Error>
where
    T: DateTimeInput,
    W: fmt::Write + ?Sized,
    DS: DateSymbols<'data>,
    TS: TimeSymbols,
{
```